package/base-files/files/lib/upgrade/nand.sh

```diff
@@ -229,22 +245,17 @@ nand_upgrade_ubinized() {
 	local ubi_file="$1"
 	local mtdnum="$(find_mtd_index "$CI_UBIPART")"
 
-	[ ! "$mtdnum" ] && {
-		CI_UBIPART="rootfs"
-		mtdnum="$(find_mtd_index "$CI_UBIPART")"
-	}
-
```

Comment:
yes, i removed this check. but AFAICT this is redundant coming from `nand_do_upgrade`, so it is not this.

Reply:
Looking at the code I believe you are right.
I guess what I observed was coincidental and not even related to your changes.
That being said, I think that `nand_upgrade_ubifs` and `nand_upgrade_tar` could share the code removing all ubiblock devices (which requrires unmounting them, if needed) and also unmounting any ubifs volumes still mounted at this point.

Reply:
@dangowrt yes i can write code to unmount and remove ubiblocks, it's trivial, but we don't know what the problem is. could you test before and after my changes, and if my changes are to blame reintroduce the `sync` and retest? either the problem preexisted my code or there is something happening that we don't understand, and i cant test it.

btw, years ago i made a gparted like tool that resized and moved around partitions, change filesystems, wiped, etc on android phones. it supported several filesystems. it could do the changes without erasing data in the affected partitions. it would lay out a plan for achieving the goal, do a dry run to test everything it could, and then proceed. if the goal still could not be reached, it would leave the device in a coherent state from which you could roll back or establish a reachable goal and roll forward. it was ported to many devices and had thousands of users and no episodes of lost info.

thing is, it was written in ash, busybox'es shell.
check it out: https://lanchon.github.io/REPIT/
and a demo run: https://raw.githubusercontent.com/Lanchon/REPIT/master/LOG.txt

so don't worry that i'll be able to handle this with a little help testing :)

Reply:
re-introducing the `sync` doesn't fix it, it's the filesystem still being mounted which causes the havoc (and that was broken already before, I wrongly blamed your change for the symptoms I saw...)
ubinized sysupgrade anyway isn't that relevant in practice and hardly used at all by OpenWrt, but as it is there it would be nice to work reliably as well.

Anyway, I'm happy to test your suggestions to fix that which would be great, but I don't have much time to participate more than couple a quick test on two devices on my desk.
And I guess I'll pick all the rest of this series after some more testing later today.

PS: I think I've even used REPIT some years ago to re-partition a Samsung Exynos-based device of a friend who was to switching to Spaget ROM....

Reply:
> I think I've even used REPIT

lol!